daemon/logs.go

```diff
@@ -125,38 +126,46 @@ func (daemon *Daemon) ContainerLogs(name string, config *ContainerLogsConfig) er
 			}
 		}
 	}
-	if config.Follow && container.IsRunning() {
-		errors := make(chan error, 2)
-		wg := sync.WaitGroup{}
-
-		if config.UseStdout {
-			wg.Add(1)
-			stdoutPipe := container.StdoutLogPipe()
-			defer stdoutPipe.Close()
-			go func() {
-				errors <- jsonlog.WriteLog(stdoutPipe, outStream, format)
-				wg.Done()
-			}()
+	if config.Follow {
+		ld := container.getLogDriver()
+		if ld == nil {
+			// container is stopped
+			return nil
 		}
-		if config.UseStderr {
-			wg.Add(1)
-			stderrPipe := container.StderrLogPipe()
-			defer stderrPipe.Close()
-			go func() {
-				errors <- jsonlog.WriteLog(stderrPipe, errStream, format)
-				wg.Done()
-			}()
+		jld, ok := ld.(*jsonfilelog.JSONFileLogger)
+		if !ok {
+			return nil
 		}
+		listener := jld.Subscribe()
 
-		wg.Wait()
-		close(errors)
-
-		for err := range errors {
-			if err != nil {
-				logrus.Errorf("%s", err)
+		defer jld.Evict(listener)
+		for msg := range listener {
+			lMsg, ok := msg.(*logger.Message)
+			if !ok {
+				// this panic is impossible :)
+				// but will be catched by mux anyway
+				panic(fmt.Sprintf("unexpected type of log message: %T", msg))
```

Comment:
It is actually should be a real panic, because of nature of `pubsub` package we're using `chan interface{}` and reading interface, so if in channel something apart from `*logger.Message` - we're in big-big-big troubles.